feat: enhance image upload validation across the application

Devanshusharma2005 opened the pull request:
## What does this PR do?

Enhancing Image upload validation. 

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Visual Demo (For contributors especially)

A visual demonstration is strongly recommended, for both the original and new change **(video / image - any one)**.

#### Video Demo (if applicable):


https://github.com/user-attachments/assets/2f519231-6541-46d7-ab39-84453aa4047c


## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

<!-- Remove bullet points below that don't apply to you -->

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

Devanshusharma2005 committed:
feat: enhance image upload validation across the application

Devanshusharma2005 committed:
Patch improvements.

Devanshusharma2005 committed:
refactor: streamline avatar upload error handling in updateProfile handler

Devanshusharma2005 committed:
refactor: extract image validation logic into a separate module for reuse in BannerUploader and ImageUploader

Devanshusharma2005 committed:
fix: reset file input value on validation failure in image uploaders

Devanshusharma2005 committed:
fix: update localization key for image file upload error message

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
fix: update HTML content validation in image uploader to check for additional byte

Devanshusharma2005 committed:
Merge branch 'fix/image-upload-validation-fix' of https://github.com/Devanshusharma2005/cal into fix/image-upload-validation-fix

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Amit91848 commented on packages/lib/server/imageValidation.ts:
```diff
@@ -0,0 +1,191 @@
+  const expectedFormat = expectedMimeType.toLowerCase().replace("image/", "");
+
+  // Handle JPEG variations
+  if (
+    (detectedFormat === "jpeg" && (expectedFormat === "jpg" || expectedFormat === "jpeg")) ||
+    (expectedFormat === "jpeg" && detectedFormat === "jpeg")
+  ) {
+    return true;
+  }
```
What are we trying to do here? Pretty confusing imo

Amit91848 commented on packages/lib/server/imageValidation.ts:
```diff
@@ -0,0 +1,191 @@
+/**
+ * Validate MIME type against actual content
+ */
+export function validateMimeType(base64Data: string, expectedMimeType: string): boolean {
```
You have created this function but haven't used it anywhere.
Can you confirm and remove?

Amit91848 commented on packages/lib/server/imageValidation.ts:
```diff
@@ -0,0 +1,191 @@
+  }
+}
+
+/**
+ * Validate file object (for frontend use)
+ */
+export async function validateImageFile(file: File): Promise<ImageValidationResult> {
```
It says validation is for frontend use so why does it exist in `/lib/server`?
Also I can see another function with same name in `/ui/components/image-uploader/imageValidation` which is actually being used in the ui components
I think the `validateImageFile` from this file is not being used anywhere, can you verify and remove this as well?

Amit91848 reviewed (changes requested):
Great work @Devanshusharma2005 , haven't tested it yet but looks promising. 
The code needs some refactoring and cleanup. Left comments

Also can you whip up a few tests for the validateBase64Image function, devin can probably help here

Devanshusharma2005 committed:
Code Quality improvements

Devanshusharma2005 committed:
add : unit tests.

Devanshusharma2005 committed:
fix : fixing some more issues.

Devanshusharma2005 committed:
fix : some import fixes

Devanshusharma2005 committed:
fix : fixing type-check issues

Devanshusharma2005 committed:
fix : some more import fixes.

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
fix : removing Duplicate max-file-size constant

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
refactor: enhance base64 validation with regex pattern

Devanshusharma2005 committed:
Merge branch 'fix/image-upload-validation-fix' of https://github.com/Devanshusharma2005/cal into fix/image-upload-validation-fix

Amit91848 reviewed (changes requested):
One minor issue with not being able to upload svg, otherwise great job 🔥 

https://github.com/user-attachments/assets/fb8c2e2a-98df-4baf-aec7-db10e0e92b4f

Devanshusharma2005 committed:
refactor: centralize MAX_IMAGE_FILE_SIZE and fix SVG checks.

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 commented on packages/ui/components/image-uploader/imageValidation.ts:
```diff
@@ -0,0 +1,78 @@
+ * This function performs comprehensive validation and shows toast messages directly
+ */
+export const validateImageFile = async (file: File, t: (key: string) => string): Promise<boolean> => {
+  if (file.size > MAX_IMAGE_FILE_SIZE) {
```
Done.

Devanshusharma2005 commented:
> One minor issue with not being able to upload svg, otherwise great job 🔥
> 
>  Screen.Recording.2025-07-31.at.6.10.00.PM.mov

fixed . SVGs are accepted now.

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Amit91848 reviewed (changes requested):
One final change @Devanshusharma2005 , looks good otherwise 🙏

Devanshusharma2005 committed:
fix : some toast fixes.

Devanshusharma2005 commented on packages/ui/components/image-uploader/imageValidation.ts:
```diff
@@ -0,0 +1,88 @@
+  if (!file.type.startsWith("image/")) {
+    showToast(t("only_image_files_allowed"), "error");
+    return false;
+  }
```
Solved.

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Amit91848 commented on packages/ui/components/image-uploader/BannerUploader.tsx:
```diff
@@ -90,14 +91,21 @@ export default function BannerUploader({
-    } else {
-      setFile(file);
+    // Validate the file and show appropriate toast message
+    const maxBannerSize = 10 * 1024 * 1024; // 10MB for banners (larger than avatars)
```
NIT: Earlier it was kept at 5MB, any reason to change?

Amit91848 commented on packages/ui/components/image-uploader/imageValidation.ts:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    // Check for dangerous file types first
+    if (matchesSignature(uint8Array, FILE_SIGNATURES.PDF)) {
+      return { isValid: false, error: "pdf_files_cannot_be_uploaded_as_images" };
+    }
+
+    if (matchesSignature(uint8Array, FILE_SIGNATURES.HTML)) {
+      return { isValid: false, error: "html_files_cannot_be_uploaded_as_images" };
+    }
+
+    if (matchesSignature(uint8Array, FILE_SIGNATURES.SCRIPT_TAG)) {
+      return { isValid: false, error: "script_files_cannot_be_uploaded_as_images" };
+    }
```
NIT: For translations could have used variables in i18n strings instead of duplicating the same message

Amit91848 reviewed (dismissed):
LGTM!

Devanshusharma2005 committed:
fixing the size of the banner.

hbjORbj reviewed (dismissed):
Would be great if you can check if something like this ensures that images files are still not selectable regardless of switching between All Files <-> Images Files

If there's no way to prevent users from selecting certain file extensions, then we will have to fully rely on the manual validation logic.

<img width="570" height="52" alt="Screenshot 2025-08-04 at 5 46 52 PM" src="https://github.com/user-attachments/assets/74f2a321-663b-4342-b2ab-304b209eb1c2" />

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
fix: update accepted image formats in BannerUploader and ImageUploader components

Devanshusharma2005 committed:
Merge branch 'fix/image-upload-validation-fix' of https://github.com/Devanshusharma2005/cal into fix/image-upload-validation-fix

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

hbjORbj commented:
@Devanshusharma2005 
> On Windows, the file picker always includes an “All Files” option, so users can still manually choose a PDF or any unsupported format, even if we specify accept="image/*"

Yeah I'm aware of it. There's this same behavior on Mac. That's why I suggested explicitly using the file extensions and testing. Lmk if it works!

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

anikdhabal committed:
Merge branch 'main' into fix/image-upload-validation-fix

Udit-takkar committed:
Merge branch 'main' into fix/image-upload-validation-fix

anikdhabal committed:
Merge branch 'main' into fix/image-upload-validation-fix

anikdhabal committed:
Merge branch 'main' into fix/image-upload-validation-fix

anikdhabal committed:
Merge branch 'main' into fix/image-upload-validation-fix

anikdhabal committed:
fix

anikdhabal commented:
@Devanshusharma2005 Mind checking the recent Coderabbit comments? Pls check it's made sense or not

Devanshusharma2005 commented:
> @Devanshusharma2005 Mind checking the recent Coderabbit comments? Pls check it's made sense or not

yeah looking into it.

Devanshusharma2005 committed:
coderabbit suggestions addressed.

anikdhabal commented on packages/ui/components/image-uploader/imageValidation.ts:
```diff
@@ -0,0 +1,123 @@
+  return data.slice(start);
+};
+
+const checkTextBasedDangerousSignature = (data: Uint8Array): string | null => {
+  const text = new TextDecoder("utf-8", { fatal: false }).decode(data).toLowerCase();
+
+  if (text.includes("<html") || matchesSignature(data, FILE_SIGNATURES.HTML)) {
+    return "html_files_cannot_be_uploaded_as_images";
+  }
+
+  if (matchesSignature(data, FILE_SIGNATURES.SCRIPT_TAG)) {
+    return "script_files_cannot_be_uploaded_as_images";
+  }
+
+  return null;
+};
```
Let's address this one also

anikdhabal replied:
You don't need to do anything. @Devanshusharma2005 will address it

anikdhabal committed:
addressed coderrabit comment

anikdhabal reviewed (dismissed):
Looks good

volnei reviewed (changes requested):
I love code, really... but I wondering if we can use a library like this one. https://www.npmjs.com/package/image-validator

What do you think?

Devanshusharma2005 committed:
refactor: implement generic i18n error messages with interpolation

- Add generic 'unsupported_file_type' translation key with {{type}} interpolation
- Replace hardcoded file type error messages with reusable i18n pattern
- Update imageValidation interfaces to support errorKey and errorParams
- Refactor server-side and client-side validation to use new pattern
- Remove individual translation keys for each file type (PDF, HTML, Script, ZIP, Executable)
- Add new translation keys for other validation errors (SVG, base64, empty data, etc.)
- Type-safe error handling with proper interpolation support

Benefits:
- Single reusable translation pattern reduces duplication
- Easier to maintain and localize
- Consistent error messaging across file types

Devanshusharma2005 committed:
feat: add MAX_BANNER_SIZE constant and update file size validation

- Add MAX_BANNER_SIZE constant (5MB) to shared constants file
- Update BannerUploader to use shared constant instead of inline value
- Update ImageUploader to handle new errorKey/errorParams pattern
- Maintain backward compatibility with existing error handling

Note: Pre-existing linting warnings in constants.ts and BannerUploader.tsx
are unrelated to these changes and were present before this refactor.

Devanshusharma2005 committed:
adressed volnei's suggestions.

Devanshusharma2005 committed:
test: update image validation tests for new errorKey/errorParams pattern

- Update all dangerous file type tests to expect errorKey and errorParams instead of hardcoded error messages
- Add comprehensive tests for new error format validation pattern
- Add tests for backward compatibility with error field
- Add tests for MAX_BANNER_SIZE constant integration
- Verify that unsupported file types now return generic 'unsupported_file_type' key with type interpolation
- Ensure all existing functionality continues to work with enhanced error handling

All 42 tests passing ✅

Devanshusharma2005 committed:
test: update server-side image validation tests for new error format

- Update all dangerous file type tests to expect errorKey and errorParams
- Change hardcoded error messages to i18n keys (unsupported_file_type, svg_contains_dangerous_content, etc.)
- Update edge case tests to use new error keys (empty_image_data, invalid_base64_format, unrecognized_image_format)
- Add comprehensive tests for new error format validation pattern
- Add tests for backward compatibility with error field
- Verify that unsupported file types return generic 'unsupported_file_type' key with type interpolation

All 26 server-side tests passing ✅
Complements client-side test updates for complete test coverage

Devanshusharma2005 committed:
Merge branch 'main' into fix/image-upload-validation-fix

Devanshusharma2005 commented on packages/lib/server/imageValidation.ts:
```diff
@@ -0,0 +1,109 @@
+        error: "Executable files cannot be uploaded as images",
+        detectedFormat: "Executable",
+      };
+    }
```
Good point. Reduced the locales.

Devanshusharma2005 commented:
> I love code, really... but I wondering if we can use a library like this one. https://www.npmjs.com/package/image-validator
> 
> What do you think?


Did I just build a validation library in our codebase from scratch. Haha 😂.
